### ISSUE TITLE
Superiority/Non-Inferiority, Equivalence, Equality Test for proportion difference, OR, RR

### ISSUE THREAD

**Issue by PharmCat** [JuliaStats/HypothesisTests.jl#217]
How Superiority/Non-Inferiority, Equivalence, Equality hypothesis for proportion difference, OR, RR  can be realized?

**Comment by PharmCat** [comment]
For example... I can make and implement some like:
```
struct BinomialEquivalenceDiffTest <: HypothesisTest
    p::Float64
    x1::Int
    n1::Int
    x2::Int
    n2::Int
end
```

or

```
struct BinomialDiffTest <: HypothesisTest
    p::Float64
    x1::Int
    n1::Int
    x2::Int
    n2::Int
    lyms
    hyptype::Symbol #Hypothisis Sup/Eq/Eqiv
end
```

or

```
struct BinomialEqivalenceTest <: HypothesisTest
    p::Float64
    x1::Int
    n1::Int
    x2::Int
    n2::Int
    paramtype::Symbol #Difference/Odd ratio/Risk ratio
end
```

What will be preferable?

**Comment by wildart** [comment]
Remember, you can always store some information as  a type parameters.

```julia
struct BinomialEqivalenceTest{Ratio, Hypothesis} <: HypothesisTest
    p::Float64
    x1::Int
    n1::Int
    x2::Int
    n2::Int
end
```

Where `Ratio` type can be `Difference`, `Odd`, `Risk`, and `Hypothesis` is whatever you needed.

**Comment by PharmCat** [comment]
> Remember, you can always store some information as a type parameters.

@wildart thank you! But what implementation is near to the philosophy of this package?

May be:

```
struct TwoBinomialTest{Ratio, Hypothesis}
    x1::Int
    n1::Int
    x2::Int
    n2::Int
end
```
can be preferable?

**Comment by wildart** [comment]
I guess `TwoSampleBinomialTest` would be a better name, as a complement to the one-sample `BinomialTest`. I was a bit hasty with type parameters. In this package, the tests' data structure type parameters used to reflect the sample data type. I'd suggest keeping any auxiliary non-numerical parameters as symbolic fields.

For example,

```julia
struct TwoSampleBinomialTest <: HypothesisTest
    x1::Int
    n1::Int
    x2::Int
    n2::Int
    δ::Float64 # test margin
    param1::Symbol # an additional parameter
end
```

After all, the test data structure will depend on your implementation. You may want use the `BinomialTest` implementation as a template.

**Comment by PharmCat** [comment]
> After all, the test data structure will depend on your implementation. You may want use the `BinomialTest` implementation as a template.

Hello! I made partial implementation, it will be better to make pull request when it will be 100% ready or I can make it now for discussion? What about validation, what requirement for it?

**Comment by wildart** [comment]
Sure, make a PR. It would be easier to comment over the implementation.

**Comment by nalimilan** [comment]
You mean that in the second case there would be no hypothesis testing? That would be problematic for this package.

Something which worries me is that AFAICT currently all tests we support correspond to a single hypothesis -- there is no argument to choose the kind of hypothesis you want to test. So we should probably have one test type for equality, and a different one for superiority.

**Comment by PharmCat** [comment]
> Something which worries me is that AFAICT currently all tests we support correspond to a single hypothesis -- there is no argument to choose the kind of hypothesis you want to test. So we should probably have one test type for equality, and a different one for superiority.

Hi! So you suggest 3 types:
```
struct BinomialEqualityTest <: HypothesisTest
    x1::Int
    n1::Int
    x2::Int
    n2::Int
    #...
end

struct BinomialSuperiorityTest <: HypothesisTest
    x1::Int
    n1::Int
    x2::Int
    n2::Int
    #...
end

struct BinomialEquivalenceTest <: HypothesisTest
    x1::Int
    n1::Int
    x2::Int
    n2::Int
    #...
end
```

What about p-value for this tests? Should it corresponds to  confidence interval method?

**Comment by nalimilan** [comment]
Yes, something like this. But actually I don't understand the difference between equality and equivalence. Could you explain it?

I wonder whether there's a parallel to make with other tests. For example, t-tests currently only allow testing for equality, but it could make sense to test for superiority. Would it make sense to use the `tail` argument to `pvalue` for this instead of separate types? Are there other tests that could use the same pattern? How do other software handle this?


> What about p-value for this tests? Should it corresponds to confidence interval method?

Sorry, could you develop?

**Comment by PharmCat** [comment]
>  How do other software handle this?

In equality(inequality) test it really simple because there is no any 'margin' and it look like any other test. In equivalence and superiority you should define 'margin', because there is most cases no sense is most cases to test superiority > 0. Equivalence also have 'margin' and it same as testing two one sided (non-infreriority/superiority) tests (TOST). 

Let see as it done in SAS (example from [here](https://marc.info/?l=sas-l&m=128931913917635&w=2)):

```
PROC FREQ DATA=tsf order=data;
TABLE group*f / nocol nopercent **riskdiff**(column=1 **noninf** **_margin=.1_**
method=wald var=sample );
TITLE1 'NonInferiority test for two proportions';
run;
```

where **riskdiff** - define the metric (difference / odd ratio / risk ratio)
**noninf** - hypothesis type (equivalence / equality / superiority)
**_margin=.1_** - margin value.

So this code will test:
H0: P1 - P2 <=  -margin
Ha: P1 - P2 >  -margin

For equivalence (change myself):

```
PROC FREQ DATA=tsf order=data;
TABLE group*f / nocol nopercent **riskdiff**(column=1 **equiv** **_margin=.1_**
method=wald var=sample );
TITLE1 'Equivalence test for two proportions';
run;
```

H0: P1 - P2 <= -margin || P1 - P2  >=  margin
Ha: P1 - P2 > -margin && P1 - P2 < -margin

>  Sorry, could you develop?

So, that is 'interval hypothesis' and we test them based on corporation value of confidence interval bound and  'hypothesis margin'. For example in non-inferiority we can get confidence interval for risk difference and that compare lower bound and non-inferiority margin, and in it LowCI > -margin, than we reject H0. In this context what does p-value means? In general  'p-value = Prob(Z > z)', where 'Z' has a standard normal distribution and 'z = d/se(d)' - but 'se(d)' depend on method that was used to calculate CI (for this we should return 'se(d)' for all CI methods). From other hand 'p-value' is a solution of equation (for non-inferiority) '0 = -margin - confint(..., level = p)[1]'    

  
More SAS examples [here](https://www.pharmasug.org/proceedings/2019/ST/PharmaSUG-2019-ST-091.pdf) and [here ](https://support.sas.com/documentation/onlinedoc/stat/131/freq.pdf).

**Comment by nalimilan** [comment]
Thanks. I'm not sure I completely got the implications of your explanations regarding the p-values, but I get the idea.

Do other software than SAS support this? I'm asking because SAS is a very peculiar language so it's not easy to reuse pattern from there. It's easier to take inspiration from Python and R for example.

You haven't addressed my second series of questions BTW.

**Comment by PharmCat** [comment]
Hi!

> Thanks. I'm not sure I completely got the implications of your explanations regarding the p-values, but I get the idea.
> 
> Do other software than SAS support this? I'm asking because SAS is a very peculiar language so it's not easy to reuse pattern from there. It's easier to take inspiration from Python and R for example.

I suppose that [NCSS: Statistical Software](https://www.ncss.com/) support this. I don't know about Python, but in R just confidence interval calculation can be used to test that hypothesis. And basically there is no need to calculate p-value at all, just CI. But [HypothesisTests.jl](https://github.com/JuliaStats/HypothesisTests.jl) has a different concept. It it possible I would prefer only CI calculation. 

> You haven't addressed my second series of questions BTW.

This?

> Yes, something like this. But actually I don't understand the difference between equality and equivalence. Could you explain it?

equality  - is a one-sided hypothesis with no margin. equivalence - two one-sided hypothesis with 2 margins (like bioequivalence).
